S4.Rmd

```diff
@@ -190,6 +217,8 @@ Note that this example uses another form of `new()`: unnnamed arguments are take
 
 S4 allows you to define a "validity" method that checks if your object is valid. You set this method by calling `setValidity()`. Here's a port of our factor validity checking to S4. A validity method should return a character vector containing a list of all of the problems. If there are no problems, it should return a character vector of length 0.
 
+<!-- FIXME: generally, one would like to discourage direct slot access, providing accessors (e.g., plain-old-functions) instead, and using these throughout, including in validateFactor -->
```

Comment:
It's not ok to use direct slot access in your own methods?

Reply:
I'm still not convinced by this line of reasoning, but it's clearly standard practice in the S4 community so I need to cover it.

Reply:
Hi,
It's all about encapsulation, a central concept in OO programming, not just with R and S4 objects. The idea is that the internal representation of an object is no user business so should be hidden behind a set of getters and setters. This might sound overkill for trivial objects because for these objects there will generally be a one-to-one mapping between slots and getters/setters. For more complicated objects one might want to make some slots read-only (by providing a getter but no setter), or completely hidden aka private (by providing no getter or setter). A typical private slot is one that contains cached data.

Another common situation is to have on-the-fly computed slots e.g. IRanges objects have `start()`, `end()`, and `width()` getters but only a `start` and a `width` slot (no `end` slot). The `end()` getter just does `start(x) + end(x) - 1`. Because people don't do direct slot access, they always use the getters/setters, they don't know about this and they don't care. We could change the IRanges internals anytime e.g. have a `start` and an `end` slot without breaking their code. We could also decide that we want to store the `start` and `width` vectors as Rle objects, or use other compression tricks, and add the `as.Rle` argument to the `start()`, `end()`, and `width()` getters, these getters would still return an integer vector by default (we would of course make `as.Rle` FALSE by default) and so user code wouldn't be affected.

Even in Martin's scenario of slot renaming, you're right that it would not be ideal to have the getter/setter of a slot not be named like the slot so ultimately one would want to also rename the slot. But at least by having a getter/setter you can manage a smooth transition (via .Deprecated/.Defunct cycles). If you've exposed the internals of your objects to the users and they've started to rely on them in their code, you're stuck.

H.

Reply:
Yes, but generally encapsulation is about encapsulating the internal details from outside the class, not about encapsulating some details within the class.

Reply:
But there is no "within" the class. There is the class, and there are methods that do things with instances of the class. Extensibility in part comes from adding behaviors to existing data types. Therefore, the representation of the class needs to be encapsulated.

Reply:
The examples I gave you are all about "encapsulating the internal details from outside the class" but maybe I don't understand what you mean exactly by this.

Reply:
@hpages I thought you were responding to the thread - the text has already been changed. I understand the argument for wrapping fields with accessors, although I personally think it's overkill in many situations.

@lawremi Hmmm, that argument is somewhat compelling, but doesn't a subclass always possess the slots of its parents? I guess maybe not if you're using a virtual class as a mixin but if you're at that level of S4, I don't think you need to read this guide 😄 

Updated text on accessors at http://adv-r.hadley.nz/s4.html#accessors. It still needs rewriting, but hopefully conveys the right basic message.

Reply:
As I said, it feels overkill when there is a one-to-one mapping between slots and getters/setters. But let me refine this.

It's actually overkill only if the objects are read-only. One argument for providing setters even for trivial objects is that @<- doesn't trigger object validation by default (it only checks that the type of the right value is compatible with the expected type for the slot). So the "provide accessor methods" policy could be considered overkill only for **read-only objects with a one-to-one mapping between slots and getters**. These objects are a very small subset of all the objects found in the Bioconductor ecosystem. Not providing setters for any other object will inevitably lead to the user breaking objects without realizing it.

But even the read-only objects with a one-to-one mapping between slots and getters will benefit from having getters the day you want to rename a slot and assure a smooth transition.

Reply:
@hadley The updated text looks good.  Yes, a subclass always possesses the slots of its parents, even a virtual parent. Accessors reduce coupling, subclassing increases it, so accessors become even more important as the class hierarchy becomes more complex.

@hpages Although there might be exceptions, it's more important to have a simple, consistent policy or recommended best practice: always provide and use accessors (unless you really know what you're doing).

Reply:
I think root cause of my discontent is that S4 does not have private slots, and the bioconductor philosophy is to hence treat all slots as private. That means that there's no lightweight way to expose a public slot.

Reply:
One possible compromise is the notion of properties. These would be would be accessed like slots, but the developer would still be able to encapsulate them, if necessary. Tengfei and I explored this with the objectProperties package for reference classes. We were mostly interested in observability, but the implicit encapsulation feature would be useful more broadly.

https://en.wikipedia.org/wiki/Property_(programming)